Add migration for print_style_recipes

reedstrm opened the pull request:
- commit id for those who like that sort of thing
  - title for human readable name of print style

Fixes up the triggers for maintaining default_print_style as well.

Note that the migration test will not pass for two reasons: we managed to deploy mismatched schema vs. migrations  - the trigger as written in schema is not valid, but was corrected in the migration. I did not write the rollback to install the invalid trigger function. And two, the new schema and the migrated schema differ in column order for the tables that we add columns to. The text diff of pg_dump output would be difficult to make aware of this difference. This is a known behavior of postgresql. I don't want to bow to this implementation detail by adding all new columns at the end: we have an SQL policy of not depending on the default column order, in any case. Also, existing production schema differs in exactly this way from a new schema, due to past (by hand) migrations.

reedstrm commented:
ok travis run was weird - python 2.7 apparently didn't find the migration to test, and python 3.5 ran (and rolled back) too many.

reedstrm commented:
Locally, the `ci_test_migrations.sh` script does the right thing (python 2.7). Here's the actual output:
```
Rollback test failed:
--- old_schema.sql	2017-10-10 09:11:38.571247690 -0500
+++ rolled_back_schema.sql	2017-10-10 09:11:46.291333576 -0500
@@ -1452,7 +1452,7 @@
       UPDATE default_print_style_recipes SET
         fileid=NEW.fileid,
         recipe_type=NEW.recipe_type,
-        NEW.revised)
+        revised=NEW.revised
         WHERE print_style=NEW.print_style AND tag=NEW.tag;
   END IF;
 
Migration test failed:
--- new_schema.sql	2017-10-10 09:11:48.431357369 -0500
+++ migrated_schema.sql	2017-10-10 09:11:45.307322635 -0500
@@ -2102,12 +2102,12 @@
 
 CREATE TABLE default_print_style_recipes (
     print_style text NOT NULL,
-    title text,
     tag text,
-    commit_id text,
     fileid integer,
     recipe_type text,
-    revised timestamp with time zone NOT NULL
+    revised timestamp with time zone NOT NULL,
+    title text,
+    commit_id text
 );
 
 
@@ -2599,12 +2599,12 @@
 
 CREATE TABLE print_style_recipes (
     print_style text NOT NULL,
-    title text,
     tag text NOT NULL,
-    commit_id text,
     fileid integer,
     recipe_type text DEFAULT 'web'::text,
-    revised timestamp with time zone DEFAULT now() NOT NULL
+    revised timestamp with time zone DEFAULT now() NOT NULL,
+    title text,
+    commit_id text
 );
 ```

pumazi commented on cnxdb/archive-sql/schema/triggers.sql:
```diff
@@ -248,54 +248,67 @@ CREATE TRIGGER ruleset_trigger
+    IF FOUND and NEW.fileid != OLD.fileid THEN
+        RAISE WARNING 'Cannot change recipe file: recipe in use';
+        RETURN NULL;
+    END IF;
```
This block is preventing a recipe from being overwritten if it is in use somewhere, correct?

reedstrm replied:
Right - similar code in the BEFORE DELETE trigger.

pumazi commented on cnxdb/archive-sql/schema/triggers.sql:
```diff
@@ -248,54 +248,67 @@ CREATE TRIGGER ruleset_trigger
 
 
-CREATE OR REPLACE FUNCTION delete_from_default_recipes() RETURNS trigger AS '
+CREATE OR REPLACE FUNCTION delete_from_default_recipes() RETURNS trigger AS $$
```
Can you add something like a doc string to these trigger functions? Looks like this function has the side-effect of inserting a new recipe.

reedstrm replied:
done

reedstrm commented:
I'm going to add the print_style_recipe data injector migration here. It will import from cnx-recipes. I don't think it should be a declared dependency, though. Should I guard the imports so that the migration doesn't fail?

pumazi commented:
Sounds like a plan! 👍

reedstrm commented:
docs added

reedstrm commented:
migration … migrated. Want to re-review?

reedstrm commented:
The added migration is deferred only. It is _not _ found by the migration test script as a step to run, but if cnx_Recipes is installed, it does run (since the test uses --run-deferred), then when rollback happens, the script only rolls back the recipe install, not the schema  changes. With the 'deferred' in there, yo can not _a priori_ know how many migrations will run. I added code to `ci_test_migrations` to get the correct count now.

reedstrm commented:
ok. I found a way to get the tests to pass. I decided to be less of a curmudgeon about order-of-columns in table, and just added them at the end, in both the schema and the migration.

To deal with the mis-deployed schema, I created a separate PR: #81 and rebased this one against that one. Once that is merged, this code should pass migration tests as well.

pumazi commented:
@reedstrm, perhaps a fix to cnx-recipes is needed. Are we sure the manifest.json is in the MANIFEST.in or setup.py package_data?
